### ISSUE TITLE
Layout block doesn't show the current active layout properly when running under Sway

### ISSUE THREAD

**Issue by asahaf** [greshake/i3status-rust#404]
I'm using Sway and I have two keyboard layouts which I switch between. The keyboard layout block always shows first (EN) layout even when the current active layout is switched to the secondary layout.

**Comment by 2m** [comment]
I am also seeing wrong output in the `keyboard_layout` block, but it is different. I have multiple layouts configured in sway:

```
input "..." {
  xkb_layout us,lt
}
```

And `keyboard_layout` block always shows `us,lt` value and it does not change when I switch layouts.

**Comment by atheriel** [comment]
Can you provide the output of `setxkbmap -query` under these conditions?

**Comment by asahaf** [comment]
This is the output of `setxkbmap -query`

`Cannot open display "default display"`

**Comment by asahaf** [comment]
I've found a workaround using custom block which its command executes this script, where `"English (US)"` is your primary layout

```
#!/usr/bin/env bash

x="`swaymsg -t get_inputs | jq --raw-output '[.[] | select(.type == "keyboard") | .xkb_active_layout_name | select(contains("English (US)") | not)] | first'`"
if [[ "${x}" == "null" ]]; then
    echo English
else
    echo "${x}"
fi
```

Credit: This solution is posted by @jeremejevs at [https://github.com/Alexays/Waybar/pull/85#issuecomment-440228399](url)

**Comment by 2m** [comment]
> Can you provide the output of setxkbmap -query under these conditions?

The output is
```
─╼ setxkbmap -query                                                             0
rules:      evdev
model:      pc105
layout:     us
```

and the output is the same independent of which layout is currently active.

`swaylock` detects it correctly, by the way.

**Comment by atheriel** [comment]
Sway seems to respond to `setxkbmap`, so I assumed that it would update the state reported by `-query`. This does not appear to be the case, and if anyone can track down why that would be appreciated.

If this is not a bug, we may need to add a new `driver` for Sway. I'm not sure if there is a standard way to retrieve the current layout on that platform, but we could use it without much difficulty if it exists. Parsing the output of `swaymsg -t get_inputs` does not seem very future-proof.

**Comment by 2m** [comment]
> And keyboard_layout block always shows us,lt value and it does not change when I switch layouts.

I noticed that I had old X11 configuration in `/etc/X11/xorg.conf.d/00-keyboard.conf` with `Option "XkbLayout" "us,lt"`. After removing that, keyboard layout now stays empty. I have block configured to use `localebus` driver.

And `busctl` confirms that `X11Layout` is not changed in sway when layout changes:

```
 ─╼ busctl get-property org.freedesktop.locale1 /org/freedesktop/locale1 org.freedesktop.locale1 X11Layout
s ""
```

**Comment by 2m** [comment]
Another datapoint: swaylock recognizes the correct keyboard layout when in the lock screen. It uses [xkbcommon](https://xkbcommon.org/) for that.

**Comment by atheriel** [comment]
There appears to be an [xkbcommon crate](https://crates.io/crates/xkbcommon), but it's not clear how to retrieve the current layout or listen for changes from the docs. If anyone is willing to take this on I'd happily accept a PR.

**Comment by vvrein** [comment]
While digging this issue, I found a post describing this issue best. 
https://who-t.blogspot.com/2016/12/the-future-of-xinput-xmodmap-setxkbmap.html
There is no one entry point for wm that uses Wayland protocol to get kblayout or something else related.
For sway it would be `swaymsg -t get_inputs` for other wm - other approach.

**Comment by vvrein** [comment]
By the way, swaylock seems understand current keyboard layout only after you start typing in it, not when it actually changes.

**Comment by asahaf** [comment]
sway 1.2 provides a way to subscribe to input change event

`swaymsg -t subscribe  '["input"]' --monitor`

now when you change the layout, you get notified

**Comment by atheriel** [comment]
It may be possible to use the `i3ipc` crate for this as well.

**Comment by ammgws** [comment]
Unfortunately it doesn't support sway's get_inputs or input events and the maintainer has been AWOL for over 6 months.

**Comment by vvrein** [comment]
And what about https://github.com/leshow/tokio-i3ipc ?
(just few-seconds-search, idk does it really could help)

**Comment by ammgws** [comment]
There's a sway one but might not be backwards compatible with i3. On mobile atm

**Comment by ammgws** [comment]
https://github.com/JayceFayne/swayipc-rs/blob/master/src/socket.rs

If only this was updated to check for I3SOCK as well, then I think it could be used for i3 as well since sway is meant to be backwards compatible wrt the JSON protocol.

**Comment by vvrein** [comment]
Found the same... seems ok for sway
![image](https://user-images.githubusercontent.com/17022100/80856174-20d2d780-8c37-11ea-9c4f-ec3bec84242d.png)

**Comment by ammgws** [comment]
I guess we could just call swaymsg ourselves

**Comment by ammgws** [comment]
@vvrein I wanted to see if we could replace i3ipc-rs that we use for the focused_window block with swayipc-rs, so I monkey patched swayipc-rs to read I3SOCK and swayipc-rs seemed to work fine on i3. So we can make this happen like this:

1. Rewrite focused_window block to use swayipc-rs and test it
2. Send PR to swayipc-rs to allow using I3SOCK when SWAYSOCK isn't available/manually
3. If maintainer accepts the PR then add a new `sway` driver to the layout block

Related: Switching to swayipc-rs will probably make #532 happen too since i3ipc-rs doesn't look like it will be updated anytime soon

**Comment by ammgws** [comment]
@asahaf @2m @vvrein 
Implemented via #670
